traffic_ops/traffic_ops_golang/login/register.go

```diff
@@ -0,0 +1,280 @@
+                     tm_user.username)
+VALUES ($1,
+        TRUE,
+        NOW(),
+        $2,
+        $3,
+        $4,
+        $4)
+RETURNING (
+	SELECT role.name
+	FROM role
+	WHERE id=tm_user.role
+) AS role,
+(
+	SELECT tenant.name
+	WHERE tenant.id=tm_user.tenant_id
+) AS tenant
+`
+
+const renewRegistrationQuery = `
+UPDATE tm_user
+SET registration_sent = now(),
+    role = $1,
+    tenant_id = $2,
+    token = $3
+WHERE email = $4
+RETURNING (
+	SELECT role.name
+	FROM role
+	WHERE id=tm_user.role
+) AS role,
+(
+	SELECT tenant.name
+	WHERE tenant.id=tm_user.tenant_id
+) AS tenant
+`
+
+var registrationEmailTemplate = template.Must(template.New("Registration Email").Parse("From: {{.From.Address.Address}}\r" + `
+To: {{.To.Address.Address}}` + "\r" + `
+Content-Type: text/html` + "\r" + `
+Subject: {{.InstanceName}} New User Registration` + "\r\n\r" + `
+<!DOCTYPE html>
+<html lang="en">
+<head>
+	<title>{{.InstanceName}} New User Registration</title>
+	<meta charset="utf-8"/>
+	<style>
+		.button_link {
+			display: block;
+			width: 130px;
+			background: #2682AF;
+			padding: 5px;
+			text-align: center;
+			border-radius: 5px;
+			color: white;
+			font-weight: bold;
+			text-decoration: none;
+			cursor: pointer;
+		}
+	</style>
+</head>
+<body>
+	<main>
+		<p>A new account has been created for you on the {{.InstanceName}} Portal. In the
+		{{.InstanceName}} Portal, you'll find a dashboard that provides access to all of your
+		Delivery Services.</p>
+		<p><a class="button_link" href="{{.RegisterURL}}?token={{.Token}}" target="_blank">Click here to finish your registration</a></p>
+	</main>
+	<footer>
+		<p>Thank you,<br/>
+		The {{.InstanceName}} Team</p>
+	</footer>
+</body>
+</html>
+`))
+
+func createRegistrationMsg(addr rfc.EmailAddress, t string, tx *sql.Tx, c config.ConfigPortal) ([]byte, error) {
+	var instanceName string
+	if err := tx.QueryRow(instanceNameQuery).Scan(&instanceName); err != nil {
+		return nil, err
+	}
+
+	var f = registrationEmailFormatter {
+		From: c.EmailFrom,
+		InstanceName: instanceName,
+		RegisterURL: c.BaseURL.String() + c.UserRegisterPath,
+		To: addr,
+		Token: t,
+	}
+
+	var tmpl bytes.Buffer
+	if err := registrationEmailTemplate.Execute(&tmpl, &f); err != nil {
+		return nil, err
+	}
+	return tmpl.Bytes(), nil
+}
+
+func RegisterUser (w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	var tx = inf.Tx.Tx
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+	defer r.Body.Close()
+
+	var req tc.UserRegistrationRequest
+	if userErr = api.Parse(r.Body, tx, req); userErr != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, userErr, nil)
+		return
+	}
+
+	if ok, err := inf.IsResourceAuthorizedToCurrentUser(int(req.TenantID)); err != nil {
+		sysErr = fmt.Errorf("Checking tenancy permissions of current user (%+v) on tenant #%d", inf.User, req.TenantID)
```

Comment:
Is there any concern that the `%+v` of the User object will expose any unintended data?

Reply:
I don't think so, it's a system log, so you'd need access to the TO server to read it. It also doesn't include any sensitive information, the `auth.CurrentUser` struct just contains id/username and permissions information.